staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go

```diff
@@ -225,36 +303,78 @@ func Start(nodeServer interface{}, opts ...Option) (result DRAPlugin, finalErr e
 		return nil, fmt.Errorf("start node client: %v", err)
 	}
 	d.plugin = plugin
-	defer func() {
-		// Clean up if we didn't finish succcessfully.
-		if r := recover(); r != nil {
-			plugin.stop()
-			panic(r)
-		}
-		if finalErr != nil {
-			plugin.stop()
-		}
-	}()
 	if !implemented {
 		return nil, errors.New("no supported DRA gRPC API is implemented and enabled")
 	}
 
 	// Now make it available to kubelet.
-	registrar, err := startRegistrar(klog.LoggerWithName(o.logger, "registrar"), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
+	registrar, err := startRegistrar(klog.NewContext(ctx, klog.LoggerWithName(logger, "registrar")), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
 	if err != nil {
 		return nil, fmt.Errorf("start registrar: %v", err)
 	}
 	d.registrar = registrar
 
+	// startGRPCServer and startRegistrar don't implement cancellation
+	// themselves, we add that for both here.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		<-ctx.Done()
+		d.plugin.stop()
+		d.registrar.stop()
+	}()
+
 	return d, nil
 }
 
+func (d *draPlugin) PublishResources(ctx context.Context, nodeResources []*resourceapi.ResourceModel) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	d.publishResources = true
+	d.nodeResources = nodeResources
+	d.updatePublishResources()
+}
+
+// updatePublishResources is called each time something changed that might
+// affect how resources are published.
+func (d *draPlugin) updatePublishResources() {
+	ctx := d.ctx
+	logger := klog.FromContext(ctx)
+	if ctx.Err() != nil {
+		// Already stopped, don't do anything.
+		return
+	}
+
+	if !d.publishResources {
+		// Nothing to publish (yet).
+		return
+	}
+
+	owner := resourceslice.Owner{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       d.nodeName,
+		// UID will be determined by the controller.
+	}
+	resources := &resourceslice.Resources{NodeResources: d.nodeResources}
+	d.resourceSliceController = resourceslice.StartController(klog.NewContext(ctx, klog.LoggerWithName(logger, "ResourceSlice controller")), d.kubeClient, d.driverName, owner, resources)
+}
+
 func (d *draPlugin) Stop() {
 	if d == nil {
 		return
 	}
+	d.cancel(errors.New("DRA plugin was stopped"))
 	d.registrar.stop()
 	d.plugin.stop()
+
+	// d.resourceSliceController is set concurrently.
+	d.mutex.Lock()
+	d.resourceSliceController.Stop()
+	d.mutex.Unlock()
```

Comment:
Should this also be moved to that goroutine in `Start()` to ensure that it happens anytime the context gets canceled for any reason.

Reply:
The `d.resourceSliceController` is using a context that is derived from the root context. Therefore cancelling that context will also stop the controller. It just doesn't block and wait for the goroutine to stop.